keps/sig-network/4444-service-routing-preference/README.md

```diff
@@ -0,0 +1,945 @@
+* **Effect:** The Kubernetes implementation will apply the specified routing
+  heuristic. It's important to note that the precise behavior of
+  implementation-specific heuristics might vary.
+
+### Notes/Constraints/Caveats (Optional)
+
+This proposal is our third attempt at an API revolving around such a
+configuration. There's a non-zero chance that we may need to revisit this again.
+
+### Risks and Mitigations
+
+* **Risk:** Having a routing preference like `Zone` comes at the risk of
+  endpoints in certain zones being overloaded if the originating traffic is
+  skewed towards a particular zone.
+
+  **Mitigation:**
+    * Emphasize in the documentation that the `Zone` preference is
+      designed for low-latency or monetory-cost reasons, with the understanding
+      that it can lead to overload within zones.
+    * Recommend approaches like having deployments per zone which can scale
+      independently of other zones.
+
+* **Risk:** Users migrating from the `service.kubernetes.io/topology-mode`
+  annotation might encounter differences in exact routing behavior:
+  * The new field doesn't support a routing preference that is exactly similar
+    to using `service.kubernetes.io/topology-mode=Auto` from the old annotation.
+  * If both field and the annotation are set, the annotation will take
+    precedence. (However, this behavior is temporary as the annotation will be
+    deprecated and removed in future releases)
+
+  **Mitigation:** Properly document the suggested migration paths with
+  limitations.
+
+## Design Details
+
+### Standard Heuristic Implementation (kube-proxy dataplane)
+
+kube-proxy (along with EndpointSlice controller, within kube-controller-manager
+as the control plane) will support the three standard routing preferences
+(`Default`, `Spread`, `Zone`).
+
+#### `Default` and `Spread`
+* Initially, kube-proxy will treat the `Default` preference the same as
+  `Spread`
+* This leverages existing implementation, requiring no major changes.
+
+#### `Zone`
+* This preference will be implemented by the use of Hints within EndpointSlices.
+* We already use Hints to implement `service.kubernetes.io/topology-mode=Auto`
+  Similarly, we’ll use the same Hints within the EndpointSlice to implement the
+  `Zone` heuristic – the hints will match the zone of the endpoint itself.
+* While it may seem redundant to populate the hints here since kube-proxy can
+  already derive the zone hint from the endpoints zone (as they would be the
+  same), we will still use this for implementation simply because of the reason
+  that it’s easier to implement and provides a better design. Consider an
+  alternative implementation where kube-proxy reads
+  `routingPreference=Zone` field and then constructs the route rules
+  accordingly. This means some extra logic needs to be baked into the kube-proxy
+  which could have just as easily been implemented by an already existing
+  extensibility mechanism (i.e. EndpointSlice hints)
```

Comment:
I don't understand this. It sounds like you are saying "having two fields that will always have the same value is better design than only having one"